docs/docs/50-user-guide/60-reference-docs/40-expressions.md

```diff
@@ -371,7 +371,78 @@ relevant `Freight` is not found from the `FreightCollection`.
 :::tip
 You can handle `nil` values gracefully in Expr using its
 [nil coalescing](https://expr-lang.org/docs/language-definition#nil-coalescing) and
-[optional chaining](https://expr-lang.org/docs/language-definition#optional-chaining) features.
+[optional chaining](https://expr-lang.org/docs/language-definition#optional-chaining)
+features.
+:::
+
+### `freightMetadata(freightName)`
+
+The `freightMetadata()` function retrieves metadata stored in a `Freight` resource.
+It has one required argument:
+
+- `freightName` (Required): The name of the `Freight` resource
+
+This returns a map containing all metadata key-value pairs stored in the `Freight`
+resource.
+
+Example:
+
+```yaml
+config:
+  # Access metadata values using ['key-name'] syntax
+  issueKey: ${{ freightMetadata(ctx.targetFreight.name)['jira-issue-key'] }}
+
+  # Using nil coalescing (??) to provide default values if metadata is missing
+  tier: ${{ stageMetadata(ctx.stage)['tier'] ?? "default-tier" }}
+
+  # Using optional chaining (?.) with nil coalescing for nested values
+  nested: ${{ stageMetadata(ctx.stage)?.config?.settings?.timeoutSeconds ?? 300 }}
+```
+
+:::tip
+You can handle `nil` values gracefully in Expr using its
+[nil coalescing](https://expr-lang.org/docs/language-definition#nil-coalescing) and
+[optional chaining](https://expr-lang.org/docs/language-definition#optional-chaining)
+features.
+:::
+
+:::note
+An optional second argument (`freightMetadata(freightName, 'key-name')`) is supported
+but deprecated as of `v1.8` and will be removed in `v1.10`. While the two-argument
+form returns a single value for the specified key, the single-argument form returns
+the complete metadata map. To migrate, use the single-argument form with map access
+syntax: `freightMetadata(freightName)['key-name']`.
+:::
+
+### `stageMetadata(stageName)`
+
+The `stageMetadata()` function retrieves metadata stored in a `Stage` resource. It
+has one required argument:
+
+- `stageName` (Required): The name of the `Stage` resource
+
+This returns a map containing all metadata key-value pairs stored in the `Stage`
+resource.
+
+Example:
+
+```yaml
+config:
+  # Access metadata values using ['key-name'] syntax
+  region: ${{ stageMetadata(ctx.stage)['aws-region'] }}
+
+  # Using nil coalescing (??) to provide default values if metadata is missing
+  tier: ${{ stageMetadata(ctx.stage)['tier'] ?? "default-tier" }}
+
+  # Using optional chaining (?.) with nil coalescing for nested values
+  nested: ${{ stageMetadata(ctx.stage)?.config?.settings?.timeoutSeconds ?? 300 }}
```

Comment:
Just a note that _this_ is a great example that addresses my worries [in this other comment](https://github.com/akuity/kargo/pull/4861/files#r2310549389).